File: C++/Problems/Problems/problem_1510.cpp
#include "problem_1510.h"

bool problem_1510::test() {
    int n = 4;

    bool expected = true;

    auto result = winnerSquareGame(n);

    return expected == result;
}

bool problem_1510::winnerSquareGame(int n) {
    std::vector dp(n + 1, -1);
    return canWin(dp, n);
}

bool problem_1510::canWin(std::vector<int> &dp, int n) {
    if (n <= 0)
        return false;
    if (dp[n] != -1)
        return dp[n] == 1;
    double limit = std::sqrt(n + 1);
    int res = 0;
    for (int i = 1; i < limit; ++i)
    {
        if (!canWin(dp, n - i * i))
        {
            res = 1;
            break;
        }
    }
    dp[n] = res;
    return res == 1;
}

File: Python/Problems/Problems/problem1510.py
import heapq
import math
import queue
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter, deque

from .problem import Problem

class Problem1510(Problem):
    def test(self) -> bool:
        n = 4

        expected = True

        result = self.winnerSquareGame(n)

        return expected == result

    def winnerSquareGame(self, n: int) -> bool:
        dp = [-1] * (n + 1)
        return self.canWin(dp, n)

    def canWin(self, dp: List[int], n: int) -> bool:
        if n <= 0:
            return False
        if dp[n] != -1:
            return dp[n] == 1
        limit = int(math.sqrt(n) + 1)
        res = 0
        for i in range(1, limit):
            if not self.canWin(dp, n - i * i):
                res = 1
                break
        dp[n] = res

        return res == 1